src/shogun/clustering/KMeansBase.cpp

```diff
@@ -0,0 +1,419 @@
+{
+	return m_use_kmeanspp;
+}
+
+void CKMeansBase::set_k(int32_t p_k)
+{
+	REQUIRE(p_k>0, "number of clusters should be > 0");
+	this->m_k=p_k;
+}
+
+int32_t CKMeansBase::get_k()
+{
+	return m_k;
+}
+
+void CKMeansBase::set_max_iter(int32_t iter)
+{
+	REQUIRE(iter>0, "number of clusters should be > 0");
+	m_max_iter=iter;
+}
+
+float64_t CKMeansBase::get_max_iter()
+{
+	return m_max_iter;
+}
+
+SGVector<float64_t> CKMeansBase::get_radiuses()
+{
+	return m_R;
+}
+
+SGMatrix<float64_t> CKMeansBase::get_cluster_centers()
+{
+	if (!m_R.vector)
+		return SGMatrix<float64_t>();
+
+	CDenseFeatures<float64_t>* lhs=
+		(CDenseFeatures<float64_t>*)distance->get_lhs();
+	SGMatrix<float64_t> centers=lhs->get_feature_matrix();
+	SG_UNREF(lhs);
+	return centers;
+}
+
+int32_t CKMeansBase::get_dimensions()
+{
+	return m_dimensions;
+}
+
+void CKMeansBase::set_fixed_centers(bool fixed)
+{
+	m_fixed_centers=fixed;
+}
+
+bool CKMeansBase::get_fixed_centers()
+{
+	return m_fixed_centers;
+}
+
+void CKMeansBase::store_model_features()
+{
+	/* set lhs of underlying distance to cluster centers */
+	CDenseFeatures<float64_t>* cluster_centers=new CDenseFeatures<float64_t>(
+			m_mus);
+
+	/* store cluster centers in lhs of distance variable */
+	CFeatures* rhs=distance->get_rhs();
+	distance->init(cluster_centers, rhs);
+	SG_UNREF(rhs);
+}
+
+SGMatrix<float64_t> CKMeansBase::kmeanspp()
+{
+	int32_t num=distance->get_num_vec_lhs();
+	SGVector<float64_t> dists=SGVector<float64_t>(num);
+	SGVector<int32_t> mu_index=SGVector<int32_t>(m_k);
+	
+	/* 1st center */
+	int32_t mu_1=CMath::random((int32_t) 0,num-1);
+	mu_index[0]=mu_1;
+	
+	/* choose a center - do k-1 times */
+	int32_t count=0;
+	while (++count<m_k)
+	{
+		float64_t sum=0.0;
+		/* for each data point find distance to nearest already chosen center */
+		for (int32_t point_idx=0;point_idx<num;point_idx++)
+		{
+			dists[point_idx]=distance->distance(mu_index[0],point_idx);
+			int32_t cent_id=1;
+
+			while (cent_id<count)
+			{
+				float64_t dist_temp=distance->distance(mu_index[cent_id],point_idx); 
+				if (dists[point_idx]>dist_temp)
+					dists[point_idx]=dist_temp; 
+				cent_id++;
+			}
+
+			dists[point_idx]*=dists[point_idx];
+			sum+=dists[point_idx];
+		}
+
+		/*random choosing - points weighted by square of distance from nearset center*/
+		int32_t mu_next=0;
+		float64_t chosen=CMath::random(0.0,sum);
+		while ((chosen-=dists[mu_next])>0)
+			mu_next++;
+
+		mu_index[count]=mu_next;
+	}
+
+	CDenseFeatures<float64_t>* lhs=(CDenseFeatures<float64_t>*)distance->get_lhs();
+	int32_t dim=lhs->get_num_features();
+	SGMatrix<float64_t> mat=SGMatrix<float64_t>(dim,m_k);
+	for (int32_t c_m=0;c_m<m_k;c_m++)
+	{
+		SGVector<float64_t> feature=lhs->get_feature_vector(c_m);
+		for (int32_t r_m=0;r_m<dim;r_m++)
+			mat(r_m,c_m)=feature[r_m];
+	}
+	SG_UNREF(lhs);
+	return mat;
+}
+
+int32_t CKMeansBase::initialize_training(CFeatures* data)
+{
+	REQUIRE(distance, "Distance not supplied.\n");
```

Comment:
`distance` is inherited from DistanceMachine which i haven't touched in this PR.